RFC-0001-native-spi.md

```diff
@@ -0,0 +1,548 @@
+![Session property flow](RFC-0002/session.png)
+
+
+
+### Functions validation and execution
+
+As the source of truth will now be whatever functions are registered in Velox, the coordinator must have a master list of these functions that is consistent with whatever is registered, otherwise functions will parse correctly according to whatever is registered in Java, yet fail during execution.  Additionally, `SHOW FUNCTIONS` should simply show functions that are eligible for execution.
+
+
+#### How functions will be known to the coordinator
+
+Velox already registers all functions for execution.  Additionally, it’s already possible to obtain a list of functions from Velox.  In order to report these functions to the coordinator, a new endpoint will be added, `/v1/functions`.  A `GET` to this endpoint will return the total list of all functions registered in Velox in JSON format.  The particular format will be compatible with `UdfFunctionSignatureMap`, an existing JSON serialization format currently used to load function definitions from files.
+
+Presto can already define custom functions with the existing SPI interface `FunctionNamespaceManager`, which facilitates the execution of SQL functions and remote UDFs.  We will enhance this integration in the Presto engine to permit the execution of remote functions during constant folding.  A new implementation will be added, `NativeFunctionNamespaceManager`.  This manager will retrieve the list of functions from the sidecar’s `/v1/functions` endpoint.
+
+Functions returned from this endpoint should behave as built-in functions to the user.  Built-in functions are currently simply functions that don’t require a catalog and schema to be specified when they are executed.  Because each `FunctionNamespaceManager` registers itself to the coordinator with a catalog and namespace to refer, we must allow the functions from the `NativeFunctionNamespaceManager` to be addressed as if they were built-in functions.  To support this, the built-in namespace will be become configurable, and a `FunctionNamespaceManager` can indicate to the coordinator that its own namespace should be considered the default namespace.  The coordinator will validate that only one such plugin has specified itself as a default namespace.  The `NativeFunctionNamespaceManager` will indicate that it is a default namespace, and hence will replace built-in functions in Java.
+
+![Function flow](RFC-0002/functions.png)
+
+
+
+#### How will constant folding work
+
+A new endpoint will be added, `/v1/functions/{function_name}`.  A `GET` to this method will return function output as a page with just one block and just one row.  It will take in parameters `arg1`, `arg2`, …, `argN` and `returnType`, all of which will be serialized `TypeSignature` values.  The C++ sidecar will interpret the type signatures and resolve a function, convert the Presto page, evaluate the function, then return the results as a Presto page and send back to the coordinator.
+
```

Comment:
Would it be possible to add your offline explanation about the 2 options and their pros and cons here? 

"from that point is no different from any other function, so I think it makes most sense to do it in the sidecar (alternative is to keep track of what functions we want to duplicate in Java and what functions we don't)"

Also, IF the performance of the simple functions, shipped to the sidecar, is a lot slower than in Java process, what can we do to improve that?

Reply:
Added, PTAL!